README.rst
********
pypi-cli
********

.. image:: https://badge.fury.io/py/pypi-cli.png
    :target: http://badge.fury.io/py/pypi-cli
    :alt: Latest version

.. image:: https://travis-ci.org/sloria/pypi-cli.png?branch=master
    :target: https://travis-ci.org/sloria/pypi-cli
    :alt: Travis-CI

A command-line interface to the Python Package Index (PyPI). Get package info, download statistics, and more.

.. image:: https://dl.dropboxusercontent.com/u/1693233/github/pypi-cli.png
    :alt: Screenshot
    :target: http://konch.readthedocs.org

Get it now
==========
::

    $ pip install pypi-cli


Requirements
============

- Python >= 2.7 or >= 3.3

Examples
========

Get Package Info
----------------

Use the ``pypi info`` command to get summary information for a package.

.. code-block:: bash

    $ pypi info matplotlib

::

    matplotlib
    ==========
    Python plotting package

    Latest release:   1.3.1

    Last day:           2,015
    Last week:         16,744
    Last month:        59,989

    Author:   John D. Hunter, Michael Droettboom
    Author email: mdroe@stsci.edu

    PyPI URL:  http://pypi.python.org/pypi/matplotlib
    Home Page: http://matplotlib.org

    License: BSD


Get Download Statistics
-----------------------

Use the ``pypi stat`` command to get download statistics for a package.

.. code-block:: bash

    $ pypi stat numpy

::

    Fetching statistics for 'http://pypi.python.org/pypi/numpy'. . .

    Download statistics for numpy
    =============================

    Downloads by version
    1.0     06/12/02 [ 1,904     ] *
    1.3.0   09/04/06 [ 34,900    ] **
    1.4.1   10/04/24 [ 17,977    ] *
    1.5.0   10/09/15 [ 23,462    ] **
    1.5.1   10/11/18 [ 49,311    ] ***
    1.6.0   11/05/14 [ 39,431    ] **
    1.6.1   11/07/24 [ 168,287   ] ********
    1.6.2   12/05/20 [ 374,288   ] ******************
    1.7.0   13/02/12 [ 147,759   ] *******
    1.7.1   13/04/07 [ 1,006,008 ] **********************************************
    1.8.0   13/10/30 [ 513,208   ] ************************
    1.7.2   13/12/31 [ 2,974     ] *
    1.8.1   14/03/26 [ 356,674   ] *****************

    Min downloads:          1,904 (1.0)
    Max downloads:      1,006,008 (1.7.1)
    Avg downloads:        210,475
    Total downloads:    2,736,183

    Last day:           3,901
    Last week:         44,842
    Last month:       182,480

Browse to a Package's PyPI or homepage
--------------------------------------

Use ``pypi browse`` to open a package's PyPI url in your browser.

.. code-block:: bash

    $ pypi browse textblob

You can also go to a package's homepage.

.. code-block:: bash

    $ pypi browse textblob --homepage

Search For Packages
-------------------

Use ``pypi search`` to search for PyPI packages.

.. code-block:: bash

    $ pypi search 'requests oauth'

::

    Search results for "requests oauth"
    suds_requests
    oauth
    requests
    requests-oauthlib
    requests-foauth
    requests-oauth
    requests-oauth2
    wsgioauth
    pmr2.oauth
    django-oauth-plus


More
====

To get help or list available commands:

.. code-block:: bash

    $ pypi --help

You can also get help with subcommands:

.. code-block:: bash

    $ pypi stat --help


License
=======

MIT licensed. See the bundled `LICENSE <https://github.com/sloria/pypi-cli/blob/master/LICENSE>`_ file for more details.


HISTORY.rst
*********
Changelog
*********

0.3.3 (2014-12-22)
==================

- Don't pin to python-dateutil==2.2.

0.3.2 (2014-09-09)
==================

- Use HTTPS urls. Thanks Alex Gaynor.
- Fix error in search when a package's summary is ``None``. Thanks Eric Lo.

0.3.1 (2014-06-18)
==================

- Disable pager for displaying search results until click has better support for ANSI styles in the pager. Thanks Arne Neumann for reporting.
- Support newer versions of click. Thanks Juraj Bubniak for reporting.

0.3.0 (2014-06-15)
==================

- Fix bug that caused search to raise a `TypeError` on Python 3. Thanks @barrio for reporting.
- Search command shows package summaries.
- Search command shows all results in the pager.


0.2.1 (2014-06-12)
==================

- Fix bug that caused a crash in search. Thanks Christian Pedersen for reporting.


0.2.0 (2014-06-09)
==================

- Add ``--web`` option to `search` command.
- Fix bug with locating packages with an underscore character.
- Add license, maintainer, and documentation info to the ``info`` command.


0.1.0 (2014-06-08)
==================

- First release.
- Supports ``stat``, ``browse``, and ``search``, and ``info`` commands.


tests/conftest.py
# -*- coding: utf-8 -*-
import os

import pytest
import responses
from click.testing import CliRunner

import pypi_cli as pypi

HERE = os.path.abspath(os.path.dirname(__file__))

@pytest.yield_fixture
def mock_api():
    """A mock for the PyPI JSON API."""
    with open(os.path.join(HERE, 'response.json'), 'r') as fp:
        webargs_response = fp.read()
    # A valid package with a proper response
    responses.add(
        responses.GET,
        'https://pypi.python.org/pypi/webargs/json',
        body=webargs_response,
        content_type='application/json'
    )
    # A valid package with no releases
    with open(os.path.join(HERE, 'response_noreleases.json'), 'r') as fp:
        foo_response = fp.read()

    responses.add(
        responses.GET,
        'https://pypi.python.org/pypi/foo/json',
        body=foo_response,
        content_type='application/json'
    )

    # An invalid package name
    responses.add(
        responses.GET,
        'https://pypi.python.org/pypi/nope/json',
        status=404
    )
    responses.start()

    yield responses

    responses.stop()


@pytest.fixture
def package():
    return pypi.Package('webargs')


@pytest.fixture(scope='function')
def runner(mock_api):
    return CliRunner()


tests/test_cli.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals
import mock
import pytest

import pypi_cli as pypi

@pytest.mark.usefixtures('mock_api')
class TestStat:

    def test_missing_package_arg(self, runner):
        result = runner.invoke(pypi.cli, ['stat'])
        assert result.exit_code > 0

    def test_with_package(self, runner):
        result = runner.invoke(pypi.cli, ['stat', 'webargs'])
        assert result.exit_code == 0
        assert 'Download statistics for webargs' in result.output

    def test_with_package_url(self, runner):
        result = runner.invoke(pypi.cli, ['stat', 'https://pypi.python.org/pypi/webargs'])
        assert result.exit_code == 0
        assert 'Download statistics for webargs' in result.output

@pytest.mark.usefixtures('mock_api')
class TestBrowse:

    def test_missing_package_arg(self, runner):
        result = runner.invoke(pypi.cli, ['browse'])
        assert result.exit_code > 0

    @mock.patch('pypi_cli.click.launch')
    def test_with_package(self, mock_launch, runner):
        result = runner.invoke(pypi.cli, ['browse', 'webargs'])
        assert result.exit_code == 0
        assert mock_launch.called is True
        mock_launch.assert_called_with('http://pypi.python.org/pypi/webargs')

@pytest.mark.usefixtures('mock_api')
class TestSearch:

    def test_missing_package_arg(self, runner):
        result = runner.invoke(pypi.cli, ['search'])
        assert result.exit_code > 0

    @mock.patch('pypi_cli.click.launch')
    def test_with_package_web(self, mock_launch, runner):
        result = runner.invoke(pypi.cli, ['search', '--web', "flask marshmallow"])
        assert result.exit_code == 0, result.output
        assert mock_launch.called is True
        mock_launch.assert_called_with('https://pypi.python.org/pypi?%3Aaction=search&term=flask%20marshmallow')

@pytest.mark.usefixtures('mock_api')
class TestInfo:

    def test_missing_package_arg(self, runner):
        result = runner.invoke(pypi.cli, ['info'])
        assert result.exit_code > 0

    def test_with_package(self, runner):
        result = runner.invoke(pypi.cli, ['info', 'webargs'])
        assert result.exit_code == 0
        assert 'webargs' in result.output

    def test_with_package_url(self, runner):
        result = runner.invoke(pypi.cli, ['info', 'https://pypi.python.org/pypi/webargs'])
        assert result.exit_code == 0

def test_version(runner):
    result = runner.invoke(pypi.cli, ['-v'])
    assert result.output == pypi.__version__ + '\n'
    result = runner.invoke(pypi.cli, ['--version'])
    assert result.output == pypi.__version__ + '\n'


tests/test_utils.py
# -*- coding: utf-8 -*-
import pypi_cli as pypi


def test_no_division_by_zero_in_bargraph():
    assert pypi.TICK not in pypi.bargraph({'foo': 0})


piptool.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""piptool
Prints download statistics for PyPI packages.

Usage:
    piptool <package> ...
"""
from __future__ import unicode_literals, division, print_function
import re
import sys
import time
import math
from collections import OrderedDict

import requests
from dateutil.parser import parse as dateparse
import click
from click import echo, style
from click.termui import get_terminal_size

__version__ = "0.3.0-dev"
__author__ = "Steven Loria"
__license__ = "MIT"

DATE_FORMAT = "%y/%m/%d"
MARGIN = 5
TICK = '*'
DEFAULT_PYPI = 'http://pypi.python.org/pypi'
PYPI_RE = re.compile('''^(?:(?P<pypi>https?://[^/]+/pypi)/)?
                        (?P<name>[-A-Za-z0-9.]+)
                        (?:/(?P<version>[-A-Za-z0-9.]+))?$''', re.X)

# String length added by bold formatting
_BOLD_LEN = 8
# String length added by color formatting
_COLOR_LEN = 9

def print_version(ctx, param, value):
    if not value or ctx.resilient_parsing:
        return
    echo(__version__)
    ctx.exit()


def echof(s, *args, **kwargs):
    echo(style(s, *args, **kwargs))


@click.group()
@click.option('--version', '-v',
    is_flag=True,
    callback=print_version,
    expose_value=False,
    is_eager=True)
def cli():
    """The piptool CLI."""
    pass

def abort_not_found(name):
    echo('No versions of "{0}" were found. Please try your search again.'.format(name), file=sys.stderr)
    sys.exit(1)


@cli.command()
@click.option('--graph/--no-graph', '-g/-q', default=True, help="Output a graph of download counts.")
@click.argument('package', nargs=-1, required=True)
def stat(package, graph):
    """Print download statistics for a package.

    Example: piptool stat requests
    """
    client = requests.Session()
    for input_name in package:
        m = PYPI_RE.match(input_name)
        if not m:
            echo('Invalid name or URL: "{name}"'.format(name=input_name),
                  file=sys.stderr)
            continue
        pypi_url = m.group('pypi') or DEFAULT_PYPI
        name = m.group('name')
        echo("Fetching statistics for '{url}'. . .".format(url=pypi_url))
        package = Package(name, pypi_url=pypi_url, client=client)
        try:
            version_downloads = package.version_downloads
        except NotFoundError:
            abort_not_found(name)
        min_ver, min_downloads = package.min_version
        max_ver, max_downloads = package.max_version
        avg_downloads = package.average_downloads
        total = package.downloads
        echo()
        echof("Download statistics for {name}".format(name=name), bold=True)
        if graph:
            echo()
            echo('Downloads by version')
            echo(package.chart())
        echo()
        echo("Min downloads:   {min_downloads:12,} ({min_ver})".format(**locals()))
        echo("Max downloads:   {max_downloads:12,} ({max_ver})".format(**locals()))
        echo("Avg downloads:   {avg_downloads:12,}".format(**locals()))
        echo("Total downloads: {total:12,}".format(**locals()))
        echo()
        echo('Last day:    {daily:12,}'.format(daily=package.downloads_last_day))
        echo('Last week:   {weekly:12,}'.format(weekly=package.downloads_last_week))
        echo('Last month:  {monthly:12,}'.format(monthly=package.downloads_last_month))
        echo()

@cli.command()
@click.argument('package', required=True)
def browse(package):
    """Browse to a package's PyPI or project homepage."""
    p = Package(package)
    try:
        url = p.package_url
    except NotFoundError:
        abort_not_found(package)
    echo('Opening PyPI page for "{0}"'.format(package))
    click.termui.launch(url)


def lazy_property(fn):
    """Decorator that makes a property lazy-evaluated."""
    attr_name = '_lazy_' + fn.__name__

    @property
    def _lazy_property(self):
        if not hasattr(self, attr_name):
            setattr(self, attr_name, fn(self))
        return getattr(self, attr_name)
    return _lazy_property


def bargraph(data, max_key_width=30):
    """Return a bar graph as a string, given a dictionary of data."""
    lines = []
    max_length = min(max(len(key) for key in data.keys()), max_key_width)
    max_val = max(data.values())
    max_val_length = max(len('{:,}'.format(val)) for val in data.values())
    term_width = get_terminal_size()[0]
    max_bar_width = term_width - MARGIN - (max_length + 3 + max_val_length + 3)
    template = "{key:{key_width}} [ {value:{val_width},d} ] {bar}"
    for key, value in data.items():
        try:
            bar = int(math.ceil(max_bar_width * value / max_val)) * TICK
        except ZeroDivisionError:
            bar = ''
        line = template.format(
            key=key[:max_length],
            value=value,
            bar=bar,
            key_width=max_length,
            val_width=max_val_length)
        lines.append(line)
    return '\n'.join(lines)


class PackageError(Exception):
    pass


class NotFoundError(PackageError):
    pass


class Package(object):

    def __init__(self, name, client=None, pypi_url=DEFAULT_PYPI):
        self.client = client or requests.Session()
        self.name = name
        self.url = '{pypi_url}/{name}/json'.format(pypi_url=pypi_url, name=name)

    @lazy_property
    def data(self):
        resp = self.client.get(self.url)
        if resp.status_code == 404:
            raise NotFoundError('Package not found')
        return resp.json()

    @lazy_property
    def versions(self):
        """Return a list of versions, sorted by release datae."""
        return [k for k, v in self.release_info]

    @lazy_property
    def version_downloads(self):
        """Return a dictionary of version:download_count pairs."""
        ret = OrderedDict()
        for release, info in self.release_info:
            download_count = sum(file_['downloads'] for file_ in info)
            ret[release] = download_count
        return ret

    @property
    def release_info(self):
        release_info = self.data['releases']
        # filter out any versions that have no releases
        filtered = [(ver, releases) for ver, releases in release_info.items()
                    if len(releases) > 0]
        # sort by first upload date of each release
        return sorted(filtered, key=lambda x: x[1][0]['upload_time'])

    @lazy_property
    def version_dates(self):
        ret = OrderedDict()
        for release, info in self.release_info:
            if info:
                upload_time = dateparse(info[0]['upload_time'])
                ret[release] = upload_time
        return ret

    def chart(self):
        data = OrderedDict()
        for version, dl_count in self.version_downloads.items():
            date = self.version_dates.get(version)
            date_formatted = ''
            if date:
                date_formatted = time.strftime(DATE_FORMAT,
                    self.version_dates[version].timetuple())
            key = "{0:16} {1}".format(
                style(version, fg='cyan'),
                date_formatted
            )
            data[key] = dl_count
        return bargraph(data, max_key_width=20 + _COLOR_LEN)

    @lazy_property
    def downloads(self):
        """Total download count."""
        return sum(self.version_downloads.values())

    @lazy_property
    def max_version(self):
        """Version with the most downloads."""
        data = self.version_downloads
        return max(data.items(), key=lambda item: item[1])

    @lazy_property
    def min_version(self):
        """Version with the fewest downloads."""
        data = self.version_downloads
        return min(data.items(), key=lambda item: item[1])

    @lazy_property
    def average_downloads(self):
        """Average number of downloads."""
        return int(self.downloads / len(self.versions))

    @property
    def downloads_last_day(self):
        return self.data['info']['downloads']['last_day']

    @property
    def downloads_last_week(self):
        return self.data['info']['downloads']['last_week']

    @property
    def downloads_last_month(self):
        return self.data['info']['downloads']['last_month']

    @property
    def package_url(self):
        return self.data['info']['package_url']

    def __repr__(self):
        return '<Package(name={0!r})>'.format(self.name)


pypi_cli.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
    pypi_cli
    ~~~~~~~~

    A command line interface to the Python Package Index.

    :copyright: (c) 2014 by Steven Loria.
    :license: MIT, see LICENSE for more details.
"""
from __future__ import unicode_literals, division, print_function
import re
import sys
import time
import textwrap
import math
from collections import OrderedDict
PY2 = int(sys.version[0]) == 2
if PY2:
    from xmlrpclib import ServerProxy
    from urllib import quote as urlquote
else:
    from xmlrpc.client import ServerProxy
    from urllib.parse import quote as urlquote

import requests
from dateutil.parser import parse as dateparse
import click
from click import echo, style, echo_via_pager
from click.termui import get_terminal_size

__version__ = '0.4.0'
__author__ = 'Steven Loria'
__license__ = 'MIT'

DATE_FORMAT = "%y/%m/%d"
MARGIN = 3
DEFAULT_SEARCH_RESULTS = 100

TICK = '*'
DEFAULT_PYPI = 'https://pypi.python.org/pypi'
PYPI_RE = re.compile('''^(?:(?P<pypi>https?://[^/]+/pypi)/)?
                        (?P<name>[-A-Za-z0-9_.]+)
                        (?:/(?P<version>[-A-Za-z0-9.]+))?$''', re.X)
SEARCH_URL = 'https://pypi.python.org/pypi?%3Aaction=search&term={query}'

# Number of characters added by bold formatting
_BOLD_LEN = 8
# Number of characters added by color formatting
_COLOR_LEN = 9


def echof(s, *args, **kwargs):
    echo(style(s, *args, **kwargs))


@click.group(context_settings={'help_option_names': ('-h', '--help')})
@click.version_option(__version__, '--version', '-v', message='%(version)s')
def cli():
    """The pypi CLI.

    \b
    Examples:
        \b
        pypi stat Django
        pypi browse Flask

    To get help with a subcommand, add the --help option after the command.

    \b
        pypi stat --help
    """
    pass


def abort_not_found(name):
    raise click.ClickException('No versions of "{0}" were found. Please try '
        'your search again. NOTE: Case matters.'.format(name))

def echo_header(text):
    echo(style(text, bold=True))
    echo(style('=' * len(text), bold=True))


def get_package(name_or_url, client=None):
    m = PYPI_RE.match(name_or_url)
    if not m:
        return None
    pypi_url = m.group('pypi') or DEFAULT_PYPI
    name = m.group('name')
    return Package(name, pypi_url=pypi_url, client=client)


@cli.command()
@click.option('--graph/--no-graph', '-g/-q', default=True,
    help="Output a graph of download counts.")
@click.argument('package', nargs=-1, required=True)
def stat(package, graph):
    """Print download statistics for a package.

    \b
    Example:

        pypi stat requests
    """
    client = requests.Session()
    for name_or_url in package:
        package = get_package(name_or_url, client)
        if not package:
            echo(style(
                'Invalid name or URL: "{name}"'.format(name=name_or_url), fg='red'),
                file=sys.stderr)
            continue
        try:
            version_downloads = package.version_downloads
        except NotFoundError:
            echo(style('No versions found for "{0}". Skipping. . .'.format(package.name),
                fg='red'), file=sys.stderr)
            continue
        echo("Fetching statistics for '{url}'. . .".format(url=package.package_url))
        min_ver, min_downloads = package.min_version
        max_ver, max_downloads = package.max_version
        if min_ver is None or max_ver is None:
            raise click.ClickException('Package has no releases')
        avg_downloads = package.average_downloads
        total = package.downloads
        echo()
        header = "Download statistics for {name}".format(name=package.name)
        echo_header(header)
        if graph:
            echo()
            echo('Downloads by version')
            echo(package.chart())
        echo()
        echo("Min downloads:   {min_downloads:12,} ({min_ver})".format(**locals()))
        echo("Max downloads:   {max_downloads:12,} ({max_ver})".format(**locals()))
        echo("Avg downloads:   {avg_downloads:12,}".format(**locals()))
        echo("Total downloads: {total:12,}".format(**locals()))
        echo()
        echo_download_summary(package)
        echo()

def echo_download_summary(package):
    echo('Last day:    {daily:12,}'.format(daily=package.downloads_last_day))
    echo('Last week:   {weekly:12,}'.format(weekly=package.downloads_last_week))
    echo('Last month:  {monthly:12,}'.format(monthly=package.downloads_last_month))


@cli.command()
@click.option('--homepage', is_flag=True, default=False)
@click.argument('package', required=True)
def browse(package, homepage):
    """Browse to a package's PyPI or project homepage."""
    p = Package(package)
    try:
        if homepage:
            echof('Opening homepage for "{0}"...'.format(package), bold=True)
            url = p.home_page
        else:
            echof('Opening PyPI page for "{0}"...'.format(package), bold=True)
            url = p.package_url
    except NotFoundError:
        abort_not_found(package)
    click.launch(url)

def format_result(result, name_column_width=25):
    name = result['name']
    summary_wrapped = textwrap.wrap(
        result['summary'] or name,
        width=click.get_terminal_size()[0] - name_column_width - MARGIN
    )
    padding = ' ' * (name_column_width + 3)
    summary = ('\n' + padding).join(summary_wrapped)
    return '{0} - {1}'.format(
        style(name.ljust(name_column_width), fg='cyan', bold=True),
        summary
    )

@cli.command()
@click.option('--web', '-w', is_flag=True, default=False,
    help='Open search results in your web browser.')
@click.option('--n-results', '-n', default=DEFAULT_SEARCH_RESULTS,
    help='Max number of results to show.')
@click.argument('query', required=True, type=str)
def search(query, n_results, web):
    """Search for a pypi package.

    \b
    Examples:
        \b
        pypi search requests
        pypi search 'requests oauth'
        pypi search requests -n 20
        pypi search 'requests toolbelt' --web

    """
    if web:
        echof('Opening search page for "{0}"...'.format(query), bold=True)
        url = SEARCH_URL.format(query=urlquote(query))
        click.launch(url)
    else:
        searcher = Searcher()
        results = searcher.search(query, n=n_results)
        first_line = style('Search results for "{0}"\n'.format(query), bold=True)
        echo_via_pager(
            first_line +
            '\n'.join([format_result(result) for result in results])
        )


@cli.command()
@click.option('--license/--no-license',
    is_flag=True, default=True, help='Show license.')
@click.option('--classifiers', '-c',
    is_flag=True, default=False, help='Show classifiers.')
@click.option('--long-description', '-L',
    is_flag=True, default=False, help='Show long description.')
@click.argument('package', nargs=-1, required=True)
def info(package, long_description, classifiers, license):
    """Get info about a package or packages.
    """
    client = requests.Session()
    for name_or_url in package:
        package = get_package(name_or_url, client)
        if not package:
            echo(style(
                'Invalid name or URL: "{name}"'.format(name=name_or_url), fg='red'),
                file=sys.stderr)
            continue

        # Name and summary
        try:
            info = package.data['info']
        except NotFoundError:
            echo(style('No versions found for "{0}". Skipping. . .'.format(package.name),
                fg='red'), file=sys.stderr)
            continue
        echo_header(name_or_url)
        if package.summary:
            echo(package.summary)

        # Version info
        echo()
        echo('Latest release:   {version:12}'.format(version=info['version']))

        # Long description
        if long_description:
            echo()
            echo(package.description)

        # Download info
        echo()
        echo_download_summary(package)

        # Author info
        echo()
        author, author_email = package.author, package.author_email
        if author:
            echo('Author:   {author:12}'.format(**locals()))
        if author_email:
            echo('Author email: {author_email:12}'.format(**locals()))

        # Maintainer info
        maintainer, maintainer_email = package.maintainer, package.maintainer_email
        if maintainer or maintainer_email:
            echo()
        if maintainer:
            echo('Maintainer:   {maintainer:12}'.format(**locals()))
        if maintainer_email:
            echo('Maintainer email: {maintainer_email:12}'.format(**locals()))

        # URLS
        echo()
        echo('PyPI URL:  {pypi_url:12}'.format(pypi_url=package.package_url))
        if package.home_page:
            echo('Home Page: {home_page:12}'.format(home_page=package.home_page))
        if package.docs_url:
            echo('Documentation: {docs_url:12}'.format(docs_url=package.docs_url))


        # Classifiers
        if classifiers:
            echo()
            echo('Classifiers: ')
            for each in info.get('classifiers', []):
                echo('\t' + each)

        if license and package.license:
            echo()
            echo('License: ', nl=False)
            # license may be just a name, e.g. 'BSD' or the full license text
            # If a new line is found in the text, print a new line
            if package.license.find('\n') >= 0 or len(package.license) > 80:
                echo()
            echo(package.license)
        echo()


# Utilities
# #########

def lazy_property(fn):
    """Decorator that makes a property lazy-evaluated."""
    attr_name = '_lazy_' + fn.__name__

    @property
    def _lazy_property(self):
        if not hasattr(self, attr_name):
            setattr(self, attr_name, fn(self))
        return getattr(self, attr_name)
    return _lazy_property

def _style_value(value):
    return style('{:,}'.format(value), fg='yellow')

def bargraph(data, max_key_width=30):
    """Return a bar graph as a string, given a dictionary of data."""
    lines = []
    max_length = min(max(len(key) for key in data.keys()), max_key_width)
    max_val = max(data.values())
    max_val_length = max(
        len(_style_value(val))
        for val in data.values())
    term_width = get_terminal_size()[0]
    max_bar_width = term_width - MARGIN - (max_length + 3 + max_val_length + 3)
    template = "{key:{key_width}} [ {value:{val_width}} ] {bar}"
    for key, value in data.items():
        try:
            bar = int(math.ceil(max_bar_width * value / max_val)) * TICK
        except ZeroDivisionError:
            bar = ''
        line = template.format(
            key=key[:max_length],
            value=_style_value(value),
            bar=bar,
            key_width=max_length,
            val_width=max_val_length
        )
        lines.append(line)
    return '\n'.join(lines)


class PackageError(Exception):
    pass


class NotFoundError(PackageError):
    pass


# API Wrapper
# ###########

class Package(object):

    def __init__(self, name, client=None, pypi_url=DEFAULT_PYPI):
        self.client = client or requests.Session()
        self.name = name
        self.url = '{pypi_url}/{name}/json'.format(pypi_url=pypi_url, name=name)

    @lazy_property
    def data(self):
        resp = self.client.get(self.url)
        if resp.status_code == 404:
            raise NotFoundError('Package not found')
        return resp.json()

    @lazy_property
    def versions(self):
        """Return a list of versions, sorted by release datae."""
        return [k for k, v in self.release_info]

    @lazy_property
    def version_downloads(self):
        """Return a dictionary of version:download_count pairs."""
        ret = OrderedDict()
        for release, info in self.release_info:
            download_count = sum(file_['downloads'] for file_ in info)
            ret[release] = download_count
        return ret

    @property
    def release_info(self):
        release_info = self.data['releases']
        # filter out any versions that have no releases
        filtered = [(ver, releases) for ver, releases in release_info.items()
                    if len(releases) > 0]
        # sort by first upload date of each release
        return sorted(filtered, key=lambda x: x[1][0]['upload_time'])

    @lazy_property
    def version_dates(self):
        ret = OrderedDict()
        for release, info in self.release_info:
            if info:
                upload_time = dateparse(info[0]['upload_time'])
                ret[release] = upload_time
        return ret

    def chart(self):
        def style_version(version):
            return style(version, fg='cyan', bold=True)

        data = OrderedDict()
        for version, dl_count in self.version_downloads.items():
            date = self.version_dates.get(version)
            date_formatted = ''
            if date:
                date_formatted = time.strftime(DATE_FORMAT,
                    self.version_dates[version].timetuple())
            key = "{0:20} {1}".format(
                style_version(version),
                date_formatted
            )
            data[key] = dl_count
        return bargraph(data, max_key_width=20 + _COLOR_LEN + _BOLD_LEN)

    @lazy_property
    def downloads(self):
        """Total download count.

        :return: A tuple of the form (version, n_downloads)
        """
        return sum(self.version_downloads.values())

    @lazy_property
    def max_version(self):
        """Version with the most downloads.

        :return: A tuple of the form (version, n_downloads)
        """
        data = self.version_downloads
        if not data:
            return None, 0
        return max(data.items(), key=lambda item: item[1])

    @lazy_property
    def min_version(self):
        """Version with the fewest downloads."""
        data = self.version_downloads
        if not data:
            return (None, 0)
        return min(data.items(), key=lambda item: item[1])

    @lazy_property
    def average_downloads(self):
        """Average number of downloads."""
        return int(self.downloads / len(self.versions))

    @property
    def author(self):
        return self.data['info'].get('author')

    @property
    def description(self):
        return self.data['info'].get('description')

    @property
    def summary(self):
        return self.data['info'].get('summary')

    @property
    def author_email(self):
        return self.data['info'].get('author_email')

    @property
    def maintainer(self):
        return self.data['info'].get('maintainer')

    @property
    def maintainer_email(self):
        return self.data['info'].get('maintainer_email')

    @property
    def license(self):
        return self.data['info'].get('license')

    @property
    def downloads_last_day(self):
        return self.data['info']['downloads']['last_day']

    @property
    def downloads_last_week(self):
        return self.data['info']['downloads']['last_week']

    @property
    def downloads_last_month(self):
        return self.data['info']['downloads']['last_month']

    @property
    def package_url(self):
        return self.data['info']['package_url']

    @property
    def home_page(self):
        return self.data['info'].get('home_page')

    @property
    def docs_url(self):
        return self.data['info'].get('docs_url')

    def __repr__(self):
        return '<Package(name={0!r})>'.format(self.name)


class Searcher(object):
    """PyPI package search wrapper that uses the PyPI's XMLRPC API.

    Search algorithm adapted from Supreet Sethi's implementation (MIT Licensed).
    https://github.com/djinn/pypi-json/blob/master/LICENSE.md
    """

    STOP_WORDS = set([
        "a", "and", "are", "as", "at", "be", "but", "by",
        "for", "if", "in", "into", "is", "it",
        "no", "not", "of", "on", "or", "such",
        "that", "the", "their", "then", "there", "these",
        "they", "this", "to", "was", "will",
    ])

    NAME_MATCH_WEIGHT = 16
    CONTAINS_NAME_MULT = 4
    NAME_IN_SUMMARY_MULT = 2

    def __init__(self, pypi_url='https://pypi.python.org/pypi', client=None):
        self.pypi_url = pypi_url
        self.client = client or ServerProxy(pypi_url)

    def score(self, tokens, record):
        score = 0
        name, summary = record['name'].lower(), record['summary']
        for token in tokens:
            qtf = 0
            if token == name:
                qtf += self.NAME_MATCH_WEIGHT
            else:
                n_name_matches = len(re.compile(token).findall(name))
                qtf += self.CONTAINS_NAME_MULT * n_name_matches
            if record['summary'] is not None:
                summary_matches = re.compile(token).findall(summary.lower())
                qtf += 2 * len(summary_matches)
            score += qtf
        return score

    def search(self, query, n=None):
        tokens = [each.strip() for each in query.strip().lower().split()
                  if each not in self.STOP_WORDS]
        results = self.client.search({'name': tokens}, 'and')
        visited = []
        nd = []
        for result in results:
            name = result['name']
            try:
                visited.index(name)
            except ValueError:
                nd.append(result)
                visited.append(name)
        ranked = [(self.score(tokens, result), result) for result in nd]
        sorted_results = sorted(ranked, reverse=True, key=lambda t: t[0])
        limited_results = sorted_results[:n] if n else sorted_results
        return (result for score, result in limited_results)
